src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs

```diff
@@ -3996,6 +4053,1247 @@ End Class
+        }
+    }
+
+    static void Main()
+    {
+        M(new C());
+    }
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (2,9): error CS8850: A positional record must have both a 'data' modifier and non-empty parameter list
+                // record C()
+                Diagnostic(ErrorCode.ERR_BadRecordDeclaration, "()").WithLocation(2, 9));
+
+            Assert.Equal(
+                "void C.Deconstruct()",
+                comp.GetMember("C.Deconstruct").ToTestDisplayString(includeNonNullable: false));
+        }
+
+        [Fact]
+        public void Deconstruct_UserDefined()
+        {
+            var source =
+@"using System;
+
+record B(int X, int Y)
+{
+    public void Deconstruct(out int X, out int Y)
+    {
+        X = this.X + 1;
+        Y = this.Y + 2;
+    }
+
+    static void M(B b)
+    {
+        switch (b)
+        {
+            case B(int x, int y):
+                Console.Write(x);
+                Console.Write(y);
+                break;
+        }
+    }
+
+    public static void Main()
+    {
+        M(new B(0, 0));
+    }
+}
+";
+            var verifier = CompileAndVerify(source, expectedOutput: "12");
+            verifier.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void Deconstruct_UserDefined_DifferentSignature_01()
+        {
+            var source =
+@"using System;
+
+record B(int X, int Y)
+{
+    public void Deconstruct(out int Z)
+    {
+        Z = X + Y;
+    }
+
+    static void M(B b)
+    {
+        switch (b)
+        {
+            case B(int x, int y):
+                Console.Write(x);
+                Console.Write(y);
+                break;
+        }
+        switch (b)
+        {
+            case B(int z):
+                Console.Write(z);
+                break;
+        }
+    }
+
+    public static void Main()
+    {
+        M(new B(1, 2));
+    }
+}
+";
+            var verifier = CompileAndVerify(source, expectedOutput: "123");
+            verifier.VerifyDiagnostics();
+
+            var expectedSymbols = new[]
+            {
+                "void B.Deconstruct(out System.Int32 Z)",
+                "void B.Deconstruct(out System.Int32 X, out System.Int32 Y)",
+            };
+            Assert.Equal(expectedSymbols, verifier.Compilation.GetMembers("B.Deconstruct").Select(s => s.ToTestDisplayString(includeNonNullable: false)));
+        }
+
+        [Fact]
+        public void Deconstruct_UserDefined_DifferentSignature_02()
+        {
+            var source =
+@"using System;
+
+record B(int X)
+{
+    public int Deconstruct(out int a) => throw null;
+
+    static void M(B b)
+    {
+        switch (b)
+        {
+            case B(int x):
+                Console.Write(x);
+                break;
+        }
+    }
+
+    public static void Main()
+    {
+        M(new B(1));
+    }
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
```

Comment:
Thanks for adding. 
I'm curious what others think, but it feels like we should produce an error on the record declaration, as we're producing a record that does not abide by the record contract.

Reply:
That seems reasonable. I think something similar should happen in `Deconstruct_UserDefined_DifferentSignature_04` as well. It feels like if we find the member already in the dictionary, then we maybe need to do an exact signature comparison and give an error if anything is different.